src/signature.c

```diff
@@ -24,198 +24,342 @@
+ * the content, then verifies the hash using the publick key against the
+ * signature.
+ *
+ * returns: true if signature was correct, false otherwise */
+static bool validate_signature(FILE *fp_data, FILE *fp_sig)
+{
+	char buffer[BUFFER_SIZE];
+	unsigned char sig_buffer[BUFFER_SIZE];
+	size_t sig_len;
+	size_t data_size;
+	size_t len = 0;
+	size_t bytes_left;
+	size_t count;
+	struct stat st;
+	int fd;
+	EVP_MD_CTX md_ctx;
+
+	sig_len = fread(sig_buffer, 1, BUFFER_SIZE, fp_sig);
+	if (sig_len == 0) {
+		fprintf(stderr, "Failed to get signature length!\n");
+		goto error;
+	}
+	fd = fileno(fp_data);
+	if (fstat(fd, &st) != 0) {
+		fprintf(stderr, "Failed to stat data file\n");
+		goto error;
+	}
+	data_size = st.st_size;
+
+	/* Initialize verify context and use sha256 algorithm internally */
+	if (!EVP_VerifyInit(&md_ctx, EVP_sha256())) {
+		goto error;
+	}
+
+	/* read all bytes from file to calculate digest using sha256 and then sign it */
+	bytes_left = data_size;
+	while (bytes_left > 0) {
+		count = (bytes_left > BUFFER_SIZE ? BUFFER_SIZE : bytes_left);
+		len = fread(buffer, 1, count, fp_data);
+		if (len != count) {
+			fprintf(stderr, "Signature check failed, read %lu bytes, expected %lu\n", len, count);
+			goto error;
 		}
-		if (X509_load_crl_file(lookup, crl_filename, X509_FILETYPE_PEM) != 1) {
-			goto err;
+		/* Hashes len bytes into the verification context */
+		if (!EVP_VerifyUpdate(&md_ctx, buffer, len)) {
+			goto error;
 		}
+		bytes_left -= len;
+	}
+
+	/* Verify data in context using pkey against signature */
+	if (EVP_VerifyFinal(&md_ctx, sig_buffer, sig_len, pkey) == 1) {
+		return true;
+	}
+
+error:
+	ERR_print_errors_fp(stderr);
+	return false;
+}
+
+/* This function makes sure the certificate is still valid by not having any
+ * compromised certificates in the chain.
+ * If there is no Certificate Revocation List (CRL) it may be that the private
+ * keys have not been compromised or the CRL has not been generated by the
+ * Certificate Authority (CA)
+ *
+ * returns: true if certificate is valid, false otherwise */
+static bool validate_certificate(void)
+{
+	FILE *fp = NULL;
+	X509_STORE *store = NULL;
+	X509_LOOKUP *lookup = NULL;
+	X509_STORE_CTX *verify_ctx = NULL;
+
+	/* CRL and Chains are not required for the current setup, but we may
+	 * implement them in the future */
+	if (!crl) {
+		printf("No certificate revocation list provided\n");
+	}
+	if (!chain) {
+		printf("No certificate chain provided\n");
+	}
+
+	/* create the cert store and set the verify callback */
+	if (!(store = X509_STORE_new())) {
+		goto error;
+	}
+
+	X509_STORE_set_verify_cb_func(store, verify_callback);
+
+	/* Add the certificates to be verified to the store */
+	if (!(lookup = X509_STORE_add_lookup(store, X509_LOOKUP_file()))) {
+		goto error;
+	}
+
+	/*  Load the our Root cert, which can be in either DER or PEM format */
+	if (!X509_load_cert_file(lookup, CERTNAME, X509_FILETYPE_PEM)) {
```

Comment:
Is the code comment up-to-date? Or does this function try DER if the cert is not in PEM format?